Vector.c

```diff
@@ -242,6 +261,56 @@ Object* Vector_remove(Vector* this, int idx) {
    }
 }
 
+Object* Vector_softRemove(Vector* this, int idx) {
+   assert(idx >= 0 && idx < this->items);
+
+   Object* removed = this->array[idx];
+   this->array[idx] = NULL;
+
+   this->dirty_count++;
+   if (this->dirty_index < 0 || idx < this->dirty_index) {
+      this->dirty_index = idx;
+   }
+
+   if (this->owner) {
+      Object_delete(removed);
+      return NULL;
+   }
+
+   return removed;
+}
+
+void Vector_compact(Vector* this) {
+   if (!Vector_isDirty(this)) {
+      return;
+   }
+
+   const int size = this->items;
+   assert(0 <= this->dirty_index && this->dirty_index < size);
+   assert(this->array[this->dirty_index] == NULL);
+
+   int idx = this->dirty_index;
+
+   /* one deletion: use memmove, which should be faster */
+   if (this->dirty_count == 1) {
+      memmove(&this->array[idx], &this->array[idx + 1], (this->items - idx) * sizeof(this->array[0]));
+   } else {
+      /* multiple deletions */
+      for (int i = this->dirty_index + 1; i < size; i++) {
```

Comment:
```suggestion
      for (int i = idx + 1; i < size; i++) {
```

You could even argue to drop the `+ 1` in the initializer, but that's purely a matter of taste (and one additional loop iteration).

Reply:
I prefer, the `+1` since it makes it clear that the value at `idx` is NULL and is correct. Without the `+1` a careful reader of this code could be left wondering: "why are they starting at an index that is known to be NULL - am I missing something?"

Reply:
Also, fixed/

Reply:
The note on `+ 1` was purely a remark; I didn't intend for it to be removed, even though you validly could.